File: project_euler/Problem044/Problem044.py
import math

class Problem044:
    def solution(self) -> None:
        pentagons, index = [], 1
        while True:
            current = index * (3 * index - 1) >> 1
            for pentagon in pentagons:
                if self.__is_pentagon(current + pentagon) and self.__is_pentagon(current - pentagon):
                    print(current - pentagon)
                    return
            pentagons.append(current)
            index += 1

    def __is_pentagon(self, value: int) -> bool:
        auxiliary = math.sqrt(value * 24 + 1)
        return abs(math.floor(auxiliary + 0.5) - auxiliary) < 0.00000001 and not (int(auxiliary) + 1) % 6

File: project_euler/Problem044/Problem044.java
package project_euler;

import java.util.ArrayList;
import java.util.List;

public class Problem044 {
    public void solution() {
        List<Integer> pentagons = new ArrayList<>();
        int index = 1;
        while (true) {
            int current = index * (3 * index - 1) >> 1;
            for (int pentagon : pentagons) {
                if (isPentagon(current + pentagon) && isPentagon(current - pentagon)) {
                    System.out.println(current - pentagon);
                    return;
                }
            }
            pentagons.add(current);
            ++index;
        }
    }

    private boolean isPentagon(int value) {
        double auxiliary = Math.sqrt(value * 24 + 1);
        return Math.abs(Math.floor(auxiliary + 0.5) - auxiliary) < 0.00000001 && 0 == ((int) auxiliary + 1) % 6;
    }
}
